src/JukeBox.cpp

```diff
@@ -30,49 +49,294 @@ bool JukeBox::start(string exeName, string configFilePath) {
                                                 size_t const nrOfEntries) {
-  return Error(ErrorCode::NotImplemented, "queryTracks is not implemented yet");
+  auto tracks = mMusicBackend->queryTracks(searchPattern, nrOfEntries);
+  return tracks;
 }
 
 TResult<QueueStatus> JukeBox::getCurrentQueues(TSessionID const &) {
-  return Error(ErrorCode::NotImplemented,
-               "getCurrentQueues is not implemented yet");
+  QueueStatus qs;
+
+  auto ret = mDataStore->getQueue(QueueType::Admin);
+  if (holds_alternative<Error>(ret))
+    return get<Error>(ret);
+  qs.adminQueue = get<Queue>(ret);
+
+  ret = mDataStore->getQueue(QueueType::Normal);
+  if (holds_alternative<Error>(ret))
+    return get<Error>(ret);
+  qs.normalQueue = get<Queue>(ret);
+
+  /* Construct current PlaybackTrack through combining of information
+   * in DataStore and Spotify */
+  auto trackStore = mDataStore->getPlayingTrack();
+  if (holds_alternative<Error>(trackStore))
+    return get<Error>(trackStore);
+
+  QueuedTrack tmp = get<QueuedTrack>(trackStore);
+  PlaybackTrack pbt;
+  pbt.addedBy = tmp.addedBy;
+  pbt.album = tmp.album;
+  pbt.artist = tmp.artist;
+  pbt.duration = tmp.duration;
+  pbt.iconUri = tmp.iconUri;
+  pbt.title = tmp.title;
+  pbt.trackId = tmp.trackId;
+
+  auto trackSpotify = mMusicBackend->getCurrentPlayback();
+  if (holds_alternative<Error>(trackSpotify))
+    return get<Error>(trackSpotify);
+
+  PlaybackTrack pbtSpotify = get<PlaybackTrack>(trackSpotify);
+  pbt.progressMs = pbtSpotify.progressMs;
+  pbt.isPlaying = pbtSpotify.isPlaying;
+
+  if (!(pbt == pbtSpotify)) {
+    string msg =
+        "Jukebox.getCurrentQueues: Inconsistency between current playback "
+        "track in Spotify and DataStore";
+    LOG(WARNING) << msg;
+    return Error(ErrorCode::InvalidValue, msg);
+  }
+
+  qs.currentTrack = pbt;
+
+  return qs;
 }
 
 TResultOpt JukeBox::addTrackToQueue(TSessionID const &sid,
                                     TTrackID const &trkid,
                                     QueueType type) {
-  return Error(ErrorCode::NotImplemented,
-               "addTrackToQueue is not implemented yet");
+  auto retUser = mDataStore->getUser(sid);
+  if (holds_alternative<Error>(retUser))
+    return get<Error>(retUser);
+  User user = get<User>(retUser);
+
+  if (type == QueueType::Admin && !user.isAdmin) {
+    LOG(WARNING) << "JukeBox.addTrackToQueue: User with session ID '" << sid
+                 << "' and nickname '" << user.Name
+                 << "' is not priviledged to add a track to the admin queue.";
+    return Error(ErrorCode::AccessDenied, "User is not an admin.");
+  }
+
+  auto query = mMusicBackend->createBaseTrack(trkid);
+  if (holds_alternative<Error>(query)) {
+    LOG(WARNING) << "JukeBox.addTrackToQueue: Could not add track for TrackID '"
+                 << trkid << "'.";
+    return get<Error>(query);
+  }
+
+  auto ret = mDataStore->addTrack(get<BaseTrack>(query), type);
```

Comment:
`addTrack` itself returns a `TResultOpt`, so you can just return its return value